### DISCUSSION TITLE
Z-wave healing might be missing something

### DISCUSSION THREAD

**Discussion by lurendrejer** [zwave-js/node-red-contrib-zwave-js#121] [Q&A]
Hi, you might be missing something in the code around healing networks.

My healing progressed as it should but ended up stuck at 90+ percent.
I can't tell if the healing process is waiting for skipped nodes or simply just hasn't reported to the GUI that it has finished.
It took well over an hour to get to this point and it has been stuck for 3-4 hours now.

![image](https://user-images.githubusercontent.com/30922656/135804299-98ca8927-37b7-47c7-999f-6a15fb062a62.png)

This is just an fyi - I haven't look into anything yet.

**Comment by marcus-j-davies** [comment]
Hey,

There should be no issues in healing, do you have battery operated devices?
if so - healing is likely waiting for them to wake?

**Reply by lurendrejer** [reply]
I do, but they wake up very frequently - since i have hooked them up to a 230v powersource.

On larger z-wave networks, some nodes will be offline for long periods of time (if it is waiting for 38, 58 and 63 it will be waiting for a few months).
If it is waiting for something - the GUI might be able to desplay what it is waiting for. Homeseer just finishes the job and gives information about what nodes wasn't optimized.

**Comment by marcus-j-davies** [comment]
do you have a Gen7 Z stick? - if so what firmware?

**Reply by lurendrejer** [reply]
no, gen5.

I'll just stop the healing process manually :)

**Comment by lurendrejer** [comment]
I didn't think that something was wrong with the healing process per se - I just thought it might be something with the logic/gui to consider.

**Comment by marcus-j-davies** [comment]
So we did show everything pending in the heal progress (not just skipped, failed)
I have users with 55+ nodes in some cases and the status went of the screen during a heal 😆 

this status also went of the UI panel real estate.

Here is the completion event - I cant see where it would have the opportunity to not report the end of the heal, other then waiting for some nodes.

https://github.com/zwave-js/node-red-contrib-zwave-js/blob/b438ec9205a5c6f4344766ce54eaea9598391592/zwave-js/zwave-js.js#L1475

if you feel upto it (and feel link editing), edit the code in your install, to show the status for all nodes, under the controller.
https://github.com/zwave-js/node-red-contrib-zwave-js/blob/b438ec9205a5c6f4344766ce54eaea9598391592/zwave-js/zwave-js.js#L1495

**Reply by marcus-j-davies** [reply]
After the heal - you do get a full report back in a payload message.

**Reply by marcus-j-davies** [reply]
coming to think about it, I rarely  see skipped nodes (never yet) - if the driver does not emit the completion event (because some were skipped)

then that indeed might be the issue.
But to know for sure - we need to expose the status for all nodes in their heal progress 😃

**Comment by marcus-j-davies** [comment]
Hey, having looked at the driver source, a skipped node should not stop a completion event. The completion event is fired after the last progress report, at which point, you will receive the full report in a message, I usually store the completion report in global context so I can review later.

I can't see any problem currently, many heals have taken place, but given you found a bug in the Associations UI I'm open to the possibility 😉

You mentioned about showing the heal progress as a whole (including pending)

Well we have 6.2.0 beta currently being trialled, and this includes status icons - we could add a heal progress icon here, if you feel this will benefit.

![image](https://user-images.githubusercontent.com/55892693/135846563-d562f28f-a3c1-47de-bc46-e9fe8d543de8.png)

**Reply by lurendrejer** [reply]
I'm fine with the way things are :)

It might be the 'failed nodes' in my case - that is creating this 'issue'.
